File: backend/src/main/java/com/speedarena/engine/CarPhysics.java
package com.speedarena.engine;

import com.speedarena.model.PlayerState;
import org.springframework.stereotype.Component;

/**
 * Car physics aligned with frontend {@code carPhysics.js} (forest / Unity-style arcade).
 */
@Component
public class CarPhysics {

    public static final double TOP_SPEED = 720.0;
    public static final double MAX_REVERSE_SPEED = 250.0;
    public static final double ACCELERATION = 300.0;
    public static final double BRAKE_DECELERATION = 500.0;
    public static final double TURN_SPEED = 5.2;
    public static final double MIN_TURN_SPEED = 10.0;
    public static final double GRIP = 4.5;
    public static final double DRIFT_THRESHOLD = 350.0;
    public static final double HANDBRAKE_GRIP = 0.6;
    public static final double HANDBRAKE_TURN_BOOST = 2.5;

    public void update(PlayerState player, boolean accelerate, boolean brake,
                       boolean turnLeft, boolean turnRight, boolean handbrake, double deltaTime) {

        deltaTime = Math.min(deltaTime, 0.1);

        double speed = player.getSpeed();
        double angle = player.getAngle();
        double vx = player.getVelocityX();
        double vy = player.getVelocityY();

        if (handbrake) {
            speed *= Math.pow(0.93, deltaTime * 60);
        } else if (accelerate) {
            speed += ACCELERATION * deltaTime;
            speed = Math.min(speed, TOP_SPEED);
        } else if (brake) {
            if (speed > 0) {
                speed = Math.max(0, speed - BRAKE_DECELERATION * deltaTime);
            } else {
                speed = Math.max(-MAX_REVERSE_SPEED, speed - ACCELERATION * 0.5 * deltaTime);
            }
        } else {
            speed *= Math.pow(0.96, deltaTime * 10);
        }

        if (Math.abs(speed) > MIN_TURN_SPEED) {
            double sFactor = Math.min(1.0, Math.abs(speed) / 128.0);
            double tSpeed = TURN_SPEED * sFactor * (handbrake ? HANDBRAKE_TURN_BOOST : 1.0);
            double dir = speed >= 0 ? 1.0 : -1.0;
            if (turnLeft) {
                angle -= tSpeed * dir * deltaTime;
            } else if (turnRight) {
                angle += tSpeed * dir * deltaTime;
            }
            angle = normalizeAngle(angle);
        }
        // car angle
        double intendedVx = Math.cos(angle) * speed;
        double intendedVy = Math.sin(angle) * speed;
        double grip = handbrake ? HANDBRAKE_GRIP : GRIP;
        double blend = 1.0 - Math.exp(-grip * deltaTime);
        vx = vx + (intendedVx - vx) * blend;
        vy = vy + (intendedVy - vy) * blend;

        player.setX(player.getX() + vx * deltaTime);
        player.setY(player.getY() + vy * deltaTime);
        player.setAngle(angle);
        player.setSpeed(speed);
        player.setVelocityX(vx);
        player.setVelocityY(vy);
        player.setServerTimestamp(System.currentTimeMillis());
    }

    public void update(PlayerState player, boolean accelerate, boolean brake,
                       boolean turnLeft, boolean turnRight, double deltaTime) {
        update(player, accelerate, brake, turnLeft, turnRight, false, deltaTime);
    }

    public void update(PlayerState player, boolean accelerate, boolean brake,
                       boolean turnLeft, boolean turnRight) {
        update(player, accelerate, brake, turnLeft, turnRight, false, 1.0 / 60.0);
    }

    public static double normalizeAngle(double angle) {
        angle = angle % (2 * Math.PI);
        if (angle < 0) {
            angle += 2 * Math.PI;
        }
        return angle;
    }

    public static double angleDifference(double from, double to) {
        double diff = normalizeAngle(to - from);
        if (diff > Math.PI) {
            diff -= 2 * Math.PI;
        }
        return diff;
    }

    public static double getSpeedPercentage(double speed) {
        return Math.min(100, Math.abs(speed) / TOP_SPEED * 100);
    }

    /** Matches JS getDisplaySpeed: km/h scale */
    public static int getDisplaySpeed(double speed) {
        return (int) Math.round(Math.abs(speed) * 0.4);
    }
}





























































































/**
 * package com.speedarena.engine;

import com.speedarena.model.PlayerState;
import org.springframework.stereotype.Component;

/**
 * Car physics aligned with frontend {@code carPhysics.js} (forest / Unity-style arcade).
 
@Component
public class CarPhysics {

    public static final double TOP_SPEED = 720.0;
    public static final double MAX_REVERSE_SPEED = 250.0;
    public static final double ACCELERATION = 300.0;
    public static final double BRAKE_DECELERATION = 500.0;
    public static final double TURN_SPEED = 5.2;
    public static final double MIN_TURN_SPEED = 10.0;
    public static final double GRIP = 4.5;
    public static final double DRIFT_THRESHOLD = 350.0;
    public static final double HANDBRAKE_GRIP = 0.6;
    public static final double HANDBRAKE_TURN_BOOST = 2.5;

    public void update(PlayerState player, boolean accelerate, boolean brake,
                       boolean turnLeft, boolean turnRight, boolean handbrake, double deltaTime) {

        deltaTime = Math.min(deltaTime, 0.1);

        double speed = player.getSpeed();
        double angle = player.getAngle();
        double vx = player.getVelocityX();
        double vy = player.getVelocityY();

        if (handbrake) {
            speed *= Math.pow(0.93, deltaTime * 60);
        } else if (accelerate) {
            speed += ACCELERATION * deltaTime;
            speed = Math.min(speed, TOP_SPEED);
        } else if (brake) {
            if (speed > 0) {
                speed = Math.max(0, speed - BRAKE_DECELERATION * deltaTime);
            } else {
                speed = Math.max(-MAX_REVERSE_SPEED, speed - ACCELERATION * 0.5 * deltaTime);
            }
        } else {
            speed *= Math.pow(0.96, deltaTime * 10);
        }

        if (Math.abs(speed) > MIN_TURN_SPEED) {
            double sFactor = Math.min(1.0, Math.abs(speed) / 128.0);
            double tSpeed = TURN_SPEED * sFactor * (handbrake ? HANDBRAKE_TURN_BOOST : 1.0);
            double dir = speed >= 0 ? 1.0 : -1.0;
            if (turnLeft) {
                angle -= tSpeed * dir * deltaTime;
            } else if (turnRight) {
                angle += tSpeed * dir * deltaTime;
            }
            angle = normalizeAngle(angle);
        }
        // car angle
        double intendedVx = Math.cos(angle) * speed;
        double intendedVy = Math.sin(angle) * speed;
        double grip = handbrake ? HANDBRAKE_GRIP : GRIP;
        double blend = 1.0 - Math.exp(-grip * deltaTime);
        vx = vx + (intendedVx - vx) * blend;
        vy = vy + (intendedVy - vy) * blend;

        player.setX(player.getX() + vx * deltaTime);
        player.setY(player.getY() + vy * deltaTime);
        player.setAngle(angle);
        player.setSpeed(speed);
        player.setVelocityX(vx);
        player.setVelocityY(vy);
        player.setServerTimestamp(System.currentTimeMillis());
    }

    public void update(PlayerState player, boolean accelerate, boolean brake,
                       boolean turnLeft, boolean turnRight, double deltaTime) {
        update(player, accelerate, brake, turnLeft, turnRight, false, deltaTime);
    }

    public void update(PlayerState player, boolean accelerate, boolean brake,
                       boolean turnLeft, boolean turnRight) {
        update(player, accelerate, brake, turnLeft, turnRight, false, 1.0 / 60.0);
    }

    public static double normalizeAngle(double angle) {
        angle = angle % (2 * Math.PI);
        if (angle < 0) {
            angle += 2 * Math.PI;
        }
        return angle;
    }

    public static double angleDifference(double from, double to) {
        double diff = normalizeAngle(to - from);
        if (diff > Math.PI) {
            diff -= 2 * Math.PI;
        }
        return diff;
    }

    public static double getSpeedPercentage(double speed) {
        return Math.min(100, Math.abs(speed) / TOP_SPEED * 100);
    }

    /** Matches JS getDisplaySpeed: km/h scale 
    public static int getDisplaySpeed(double speed) {
        return (int) Math.round(Math.abs(speed) * 0.4);
    }
}

 */

File: frontend/src/game/carPhysics.js
/**
 * carPhysics.js - Full Forest Track Physics Engine (Corrected Scale)
 */

import {
  FOREST_SCALE,
  isOnRoad as isOnRoadObb,
  updateForestLapTracking,
  getStartPositionsFromLayout,
} from './track/trackGeometry.js';

export { FOREST_SCALE, getStartPositionsFromLayout } from './track/trackGeometry.js';

// ─── Physics Constants ──────────────────────────────────────────────────────

export const TOP_SPEED = 720.0;
export const MAX_REVERSE_SPEED = 250.0;
export const ACCELERATION = 300.0;
export const BRAKE_DECELERATION = 500.0;
export const FRICTION = 2.0;
export const TURN_SPEED = 5.2;
export const MIN_TURN_SPEED = 10.0;
export const GRIP = 4.5;
export const DRIFT_THRESHOLD = 350.0;

export const HANDBRAKE_GRIP = 0.6;
export const HANDBRAKE_DECELERATION = 600.0;
export const HANDBRAKE_TURN_BOOST = 2.5;

export const MAX_HEALTH = 100;

export const STARTING_POSITIONS = [
  { x: 1310, y: 400, angle: Math.PI / 2 },
  { x: 1646, y: 400, angle: Math.PI / 2 },
  { x: 1308, y: 742, angle: Math.PI / 2 },
  { x: 1644, y: 742, angle: Math.PI / 2 },
  { x: 1310, y: 572, angle: Math.PI / 2 },
  { x: 1646, y: 572, angle: Math.PI / 2 },
];

export const LAP_CONFIG = {
  totalLaps: 3,
  detectionRadius: 40 * FOREST_SCALE,
};

export function isNearPoint(carX, carY, pointX, pointY, radius) {
  const dx = carX - pointX;
  const dy = carY - pointY;
  return dx * dx + dy * dy <= radius * radius;
}

export function updateLapTracking(carState, totalLaps = 3, trackLayout = []) {
  return updateForestLapTracking(carState, totalLaps, trackLayout);
}

export function updateCarPhysics(carState, input, deltaTime) {
  deltaTime = Math.min(deltaTime, 0.1);
  let { speed, angle, velocityX = 0, velocityY = 0 } = carState;
  const { accelerate, brake, turnLeft, turnRight, handbrake = false } = input;

  if (handbrake) {
    speed *= Math.pow(0.93, deltaTime * 60);
  } else if (accelerate) {
    speed += ACCELERATION * deltaTime;
    speed = Math.min(speed, TOP_SPEED);
  } else if (brake) {
    if (speed > 0) speed = Math.max(0, speed - BRAKE_DECELERATION * deltaTime);
    else speed = Math.max(-MAX_REVERSE_SPEED, speed - ACCELERATION * 0.5 * deltaTime);
  } else {
    speed *= Math.pow(0.96, deltaTime * 10);
  }

  if (Math.abs(speed) > MIN_TURN_SPEED) {
    const sFactor = Math.min(1.0, Math.abs(speed) / 128.0);
    const tSpeed = TURN_SPEED * sFactor * (handbrake ? HANDBRAKE_TURN_BOOST : 1.0);
    const dir = speed >= 0 ? 1 : -1;
    if (turnLeft) angle -= tSpeed * dir * deltaTime;
    if (turnRight) angle += tSpeed * dir * deltaTime;
  }

  const iVx = Math.cos(angle) * speed;
  const iVy = Math.sin(angle) * speed;
  const grip = handbrake ? HANDBRAKE_GRIP : GRIP;
  const blend = 1.0 - Math.exp(-grip * deltaTime);

  carState.velocityX = (velocityX || 0) + (iVx - (velocityX || 0)) * blend;
  carState.velocityY = (velocityY || 0) + (iVy - (velocityY || 0)) * blend;
  carState.x += carState.velocityX * deltaTime;
  carState.y += carState.velocityY * deltaTime;
  carState.angle = normalizeAngle(angle);
  carState.speed = speed;
}

export function isOnRoad(x, y, trackLayout = []) {
  return isOnRoadObb(x, y, trackLayout);
}

export function handleCollision(carState, trackLayout = []) {
  if (isOnRoad(carState.x, carState.y, trackLayout)) return { collided: false, damage: 0 };
  carState.speed *= Math.pow(0.9, 1);
  return { collided: true, damage: 0.15 };
}

export function applyDamage(carState, damage) {
  carState.health = Math.max(0, (carState.health || MAX_HEALTH) - damage);
  return carState.health <= 0;
}

export function respawnCar(carState, trackLayout = []) {
  const slots = getStartPositionsFromLayout(trackLayout);
  const start = slots[0] || STARTING_POSITIONS[0];
  carState.x = start.x;
  carState.y = start.y;
  carState.angle = start.angle;
  carState.speed = 0;
  carState.velocityX = 0;
  carState.velocityY = 0;
  carState.health = MAX_HEALTH;
}

export function normalizeAngle(angle) {
  angle = angle % (2 * Math.PI);
  if (angle < 0) angle += 2 * Math.PI;
  return angle;
}

export function getDisplaySpeed(speed) {
  return Math.round(Math.abs(speed) * 0.4);
}

export function getSpeedPercentage(speed) {
  return Math.min(100, (Math.abs(speed) / TOP_SPEED) * 100);
}

/**
 * @param {object|null} spawn optional { x, y, angle } world pixels + radians
 */
export function createCarState(playerId, spawn = null) {
  let x;
  let y;
  let angle;
  if (spawn && typeof spawn.x === 'number' && typeof spawn.y === 'number' && typeof spawn.angle === 'number') {
    x = spawn.x;
    y = spawn.y;
    angle = spawn.angle;
  } else {
    const s = STARTING_POSITIONS[Math.floor(Math.random() * STARTING_POSITIONS.length)];
    x = s.x;
    y = s.y;
    angle = s.angle;
  }
  return {
    id: playerId,
    x,
    y,
    angle,
    speed: 0,
    velocityX: 0,
    velocityY: 0,
    currentLap: 0,
    isFinished: false,
    health: MAX_HEALTH,
    passedCheckpoint: false,
    lapNextSectorId: 1,
    lapFinishCooldown: false,
  };
}

export function interpolateCarState(from, to, t) {
  t = Math.max(0, Math.min(1, t));
  let aDiff = to.angle - from.angle;
  if (aDiff > Math.PI) aDiff -= 2 * Math.PI;
  if (aDiff < -Math.PI) aDiff += 2 * Math.PI;
  return {
    ...to,
    x: from.x + (to.x - from.x) * t,
    y: from.y + (to.y - from.y) * t,
    angle: normalizeAngle(from.angle + aDiff * t),
    speed: from.speed + (to.speed - from.speed) * t,
  };
}

export default {
  updateCarPhysics,
  handleCollision,
  applyDamage,
  respawnCar,
  normalizeAngle,
  getDisplaySpeed,
  getSpeedPercentage,
  createCarState,
  interpolateCarState,
  TOP_SPEED,
  MAX_HEALTH,
  STARTING_POSITIONS,
  FOREST_SCALE,
  updateLapTracking,
  LAP_CONFIG,
  isOnRoad,
  getStartPositionsFromLayout,
};
